src/main/java/week6/ParkingLot.java

```diff
@@ -0,0 +1,152 @@
+   *
+   * Assumptions:
+   *    1) len(source) = len(target)
+   *    2) arrays contain O...n, this means no duplications
+   *
+   * @param source : the initial configuration of the cars
+   * @param target : the target configuration of the cars
+   * @return : List of Moves to be performed to transform source -> target
+   */
+  public static List<Move> rearrangeCars(final int[] source, final int[] target) {
+
+    List<Move> moves = new ArrayList<Move>() {
+      @Override
+      public String toString() {
+        return super.toString().replaceAll("\\[|]|(, )", "");
+      }
+    };
+
+    if (source == null || target == null || source.length != target.length) {
+      return moves;
+    }
+
+    //creates map (wrongly positioned) car id -> position and get index of the empty Box
+    HashMap<Integer, Integer> misplacedCars = new HashMap<>();
+    int emptyBox = populateMap(source, target, misplacedCars);
+
+    while (!misplacedCars.isEmpty()) {
+      //if emptyBox is in right pos, we swap the emptyBox with one of the wrong cars
+      if (target[emptyBox] == EMPTY) {
+        emptyBox = correctEmpty(misplacedCars, target, emptyBox, moves);
+      }
+      emptyBox = moveEmpty(misplacedCars, target, emptyBox, moves);
+    }
+
+    logger.info(moves.toString());
+    return moves;
+  }
+
+
+  /*
+   * Updates the map misPlacedCars : cars that needs to be moved -> current position.   *
+   * Returns the index of the empty box
+   *
+   * Time Complexity: O(n) where n is the size of array source
+   * Space Complexity: O(1)
+   *
+   * @param source : the initial configuration of the cars
+   * @param target : the target configuration of the cars
+   * @param misplacedCars : the map to be updated cars to be moved -> their current position
+   * @return : the index of the empty box
+   */
+  private static int populateMap(int[] source, int[] target,
+      HashMap<Integer, Integer> misplacedCars) {
+    int emptyBox = -1;
+    for (int i = 0; i < source.length; i++) {
+      if (source[i] == EMPTY) {
+        emptyBox = i;
+      } else if (source[i] != target[i]) {
+        misplacedCars.put(source[i], i);
+      }
+    }
+    return emptyBox;
+  }
+
+  /*
+  * This steps is required if emptyBox is in the right position but the list is not sorted.
+  *     e.g. cars = [1, 0, 2, 3], target = [3, 0, 2, 1] and emptyBox = 1
+  * We swap the emptyBox with one of the cars in the map (wrongly positioned), here we perform
+  *     "move from 1 to 0"
+  *
+  * Amortized Time Complexity: O(1)
+  * Space Complexity: O(1)
+  *
+  * @param misplacedCars : map to be updated with the cars to be moved
+  * @param target : the target configuration of the cars
+  * @param emptyBox : the index of the emptyBox
+  * @param moves : the list of moves which is appended if emptyBox needs to be swapped
+  * @return : the index of the emptyBox
+  */
+  private static int correctEmpty(Map<Integer, Integer> misplacedCars, int[] target, int emptyBox,
```

Comment:
I think the method name could be improved. It's not really correcting anything - it's moving a car from the wrong place to the wrong place, and moving the empty slot to the wrong place! "moveSomeMisplacedCar" or "moveAnyMisplacedCar", perhaps?